bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/KeyValueStorageRocksDB.java

```diff
@@ -0,0 +1,541 @@
+            writeBatch.remove(key);
+        }
+
+        @Override
+        public void clear() {
+            writeBatch.clear();
+        }
+
+        @Override
+        public void deleteRange(byte[] beginKey, byte[] endKey) {
+            writeBatch.deleteRange(beginKey, endKey);
+        }
+
+        @Override
+        public void flush() throws IOException {
+            try {
+                db.write(optionSync, writeBatch);
+            } catch (RocksDBException e) {
+                throw new IOException("Failed to flush RocksDB batch", e);
+            }
+        }
+    }
+
+    private static final class EntryWrapper implements Entry<byte[], byte[]> {
+        // This is not final since the iterator will reuse the same EntryWrapper
+        // instance at each step
+        private byte[] key;
+        private byte[] value;
+
+        public EntryWrapper() {
+            this.key = null;
+            this.value = null;
+        }
+
+        public EntryWrapper(byte[] key, byte[] value) {
+            this.key = key;
+            this.value = value;
+        }
+
+        @Override
+        public byte[] setValue(byte[] value) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public byte[] getValue() {
+            return value;
+        }
+
+        @Override
+        public byte[] getKey() {
+            return key;
+        }
+    }
+
+    /**
+     * Checks whether the DB was already created with RocksDB, otherwise copy all
+     * data from one database into a fresh database.
+     *
+     * <p>Useful when switching from LevelDB to RocksDB since it lets us to restart
+     * fresh and build a DB with the new settings (SSTs sizes, levels, etc..)
+     */
+    private static void doUpgradeIfNeeded(String path, DbConfigType dbConfigType, ServerConfiguration conf)
+            throws IOException {
+        FileSystem fileSystem = FileSystems.getDefault();
+        final Path rocksDbMarkerFile = fileSystem.getPath(path, ROCKSDB_MARKER);
+
+        if (Files.exists(fileSystem.getPath(path))) {
+            // Database already existing
+            if (Files.exists(rocksDbMarkerFile)) {
+                // Database was already created with RocksDB
+                return;
+            }
+        } else {
+            // Database not existing, no need to convert
+            return;
+        }
+
+        log.info("Converting existing database to RocksDB: {}", path);
+        long startTime = System.nanoTime();
+
+        String rocksDbPath = path + ".rocksdb";
+        KeyValueStorage source = new KeyValueStorageRocksDB(path, dbConfigType, conf, true /* read-only */);
+
+        log.info("Opened existing db, starting copy");
+        KeyValueStorage target = new KeyValueStorageRocksDB(rocksDbPath, dbConfigType, conf);
+
+        // Copy into new database. Write in batches to speed up the insertion
+        CloseableIterator<Entry<byte[], byte[]>> iterator = source.iterator();
```

Comment:
This may fail to be closed?

Reply:
the iterator is closed in the `finally` block. In any case, this part of code was only used in a old migration at Yahoo (older code was using a different format for indexes, or was using LevelDb back in the day, so the in-place conversion was happening as needed). This will not be needed anymore.